File: tensorflow/contrib/correlation_cost/ops/correlation_cost_op.cc
// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// =============================================================================

#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"

namespace tensorflow {

using ::tensorflow::shape_inference::InferenceContext;
using ::tensorflow::shape_inference::ShapeHandle;

// --------------------------------------------------------------------------

REGISTER_OP("CorrelationCost")
    .Input("input_a: T")
    .Input("input_b: T")
    .Output("output: T")
    .Attr("kernel_size: int")
    .Attr("max_displacement: int")
    .Attr("stride_1: int")
    .Attr("stride_2: int")
    .Attr("pad: int")
    .Attr("data_format: {'NHWC', 'NCHW'} = 'NHWC'")
    .Attr("T: realnumbertype")
    .SetShapeFn([](InferenceContext* c) {
      ShapeHandle input_a, input_b, input;

      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 4, &input_a));
      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 4, &input_b));
      TF_RETURN_IF_ERROR(c->Merge(input_a, input_b, &input));

      // get input shapes
      int32 B, H, W;
      B = c->Value(c->Dim(input, 0));
      string data_format;
      Status s = c->GetAttr("data_format", &data_format);
      if (s.ok() && data_format == "NCHW") {
        H = c->Value(c->Dim(input, 2));
        W = c->Value(c->Dim(input, 3));
      } else {
        H = c->Value(c->Dim(input, 1));
        W = c->Value(c->Dim(input, 2));
      }

      int32 kernel_size;
      int32 max_displacement;
      int32 stride_1;
      int32 stride_2;
      int32 pad;

      TF_RETURN_IF_ERROR(c->GetAttr("kernel_size", &kernel_size));
      TF_RETURN_IF_ERROR(c->GetAttr("max_displacement", &max_displacement));
      // stride in input
      TF_RETURN_IF_ERROR(c->GetAttr("stride_1", &stride_1));
      // stride in patch
      TF_RETURN_IF_ERROR(c->GetAttr("stride_2", &stride_2));
      TF_RETURN_IF_ERROR(c->GetAttr("pad", &pad));

      // output channels are d**2 where, d = 2r + 1
      const int32 r = max_displacement / stride_2;
      const int32 d = 2 * r + 1;
      const int32 border = max_displacement + (kernel_size - 1) / 2;

      const int32 Cout = d * d;
      // for spatial dimensions, we pad the inputs
      const int32 Hout = static_cast<int>(
          ceil(static_cast<float>(((H + 2 * pad) - border * 2)) /
               static_cast<float>(stride_1)));
      const int32 Wout = static_cast<int>(
          ceil(static_cast<float>(((W + 2 * pad) - border * 2)) /
               static_cast<float>(stride_1)));

      // Note, the output is always NCHW (even when input is NHWC)
      c->set_output(0, c->MakeShape({B, Cout, Hout, Wout}));
      return Status::OK();
    })
    .Doc(R"doc(CorrelationCost op.)doc");

REGISTER_OP("CorrelationCostGrad")
    .Input("orig_input_a: T")
    .Input("orig_input_b: T")
    .Input("top_diff: T")
    .Output("bottom_diff_a: T")
    .Output("bottom_diff_b: T")
    .Attr("T: realnumbertype")
    .Attr("kernel_size: int")
    .Attr("max_displacement: int")
    .Attr("stride_1: int")
    .Attr("stride_2: int")
    .Attr("pad: int")
    .Attr("data_format: {'NHWC', 'NCHW'} = 'NHWC'")
    .SetShapeFn([](InferenceContext* c) {
      ShapeHandle shp_hnd;
      TF_RETURN_IF_ERROR(c->Merge(c->input(0), c->input(1), &shp_hnd));
      c->set_output(0, shp_hnd);
      c->set_output(1, shp_hnd);
      return Status::OK();
    })
    .Doc(R"doc(CorrelationCost op.)doc");

}  // namespace tensorflow

File: tensorflow/contrib/correlation_cost/python/ops/correlation_cost_op.py
# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Tensorflow op performing correlation cost operation."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from tensorflow.contrib.correlation_cost.ops import gen_correlation_cost_op
from tensorflow.contrib.util import loader
from tensorflow.python.framework import ops
from tensorflow.python.ops import array_ops
from tensorflow.python.platform import resource_loader

_correlation_cost_op_so = loader.load_op_library(
    resource_loader.get_path_to_datafile("_correlation_cost_op.so"))

# pylint: disable=redefined-builtin
def correlation_cost(input_a,
                     input_b,
                     kernel_size,
                     max_displacement,
                     stride_1,
                     stride_2,
                     pad,
                     data_format='NHWC',
                     name=None):
  """Correlation Cost Volume computation.

  Computes a cost volume using correlation for two inputs. For feature
  maps A, B with spatial dimensions w, h, c it computes

    output(a, b) = sum_{l in [-k,k]**2}  < I(a+l), J(b+l) >

  where the patches of size K=2d + 1 are centered in position a resp. b.

  The output shape is [B, C', H', W'], where

    r = max_displacement / stride_2;
    bd = max_displacement + (kernel_size - 1) / 2
    C' = (2 * r + 1) ** 2
    H' = H + 2 * (pad - bd) / stride_1
    W' = W + 2 * (pad - bd) / stride_1

  Note: When the data_format requests "NHWC", an additional explicit
    transpose operation is executed.

  Args:
    input_a: A `Tensor` of the format specified by `data_format`.
    input_b: A `Tensor` of the format specified by `data_format`.
    kernel_size: An integer specifying the height and width of the
        patch used to compute the per-patch costs.
    max_displacement: An integer specifying the maximum search radius
        for each position.
    stride_1: An integer specifying the stride length in the input.
    stride_2: An integer specifying the stride length in the patch.
    pad: An integer specifying the paddings in height and width.
    data_format: Specifies the data format.
        Possible values are:
        "NHWC" float [batch, height, width, channels]
        "NCHW" float [batch, channels, height, width]
        Defaults to `"NHWC"`.
    name: A name for the operation (optional).

  Returns:
    A `Tensor` of the format specified by `data_format`.
  """

  with ops.name_scope(name, "correlation_cost"):
    op_call = gen_correlation_cost_op.correlation_cost
    ret = op_call(input_a, input_b,
                  kernel_size=kernel_size,
                  max_displacement=max_displacement,
                  stride_1=stride_1,
                  stride_2=stride_2,
                  pad=pad,
                  data_format=data_format)
    if data_format == 'NHWC':
      # this is easier to maintain without
      # specializing an additional cuda kernel
      return array_ops.transpose(ret, [0, 2, 3, 1])
    return ret


correlation_cost_grad = gen_correlation_cost_op.correlation_cost_grad

@ops.RegisterGradient("CorrelationCost")
def _correlation_cost_grad(op, grad_output):
  kernel_size = op.get_attr("kernel_size")
  max_displacement = op.get_attr("max_displacement")
  stride_1 = op.get_attr("stride_1")
  stride_2 = op.get_attr("stride_2")
  pad = op.get_attr("pad")
  data_format = op.get_attr("data_format")

  input_a = ops.convert_to_tensor(op.inputs[0], name="input_a")
  input_b = ops.convert_to_tensor(op.inputs[1], name="input_b")
  grad_output_tensor = ops.convert_to_tensor(grad_output, name="grad_output")

  op_call = gen_correlation_cost_op.correlation_cost_grad
  grads = op_call(input_a, input_b, grad_output_tensor,
                  kernel_size=kernel_size,
                  max_displacement=max_displacement,
                  stride_1=stride_1,
                  stride_2=stride_2,
                  pad=pad,
                  data_format=data_format)


  grad_input_a = ops.convert_to_tensor(grads[0], name="grad_input_a")
  grad_input_b = ops.convert_to_tensor(grads[1], name="grad_input_b")
  return [grad_input_a, grad_input_b]
